File: prototype/rails/src/app/actions/orders/roll_up_order_status_test.rb
require "commerce_test_case"

module Orders
  class RollUpOrderStatusTest < CommerceTestCase
    def test_an_order_whose_fulfillments_all_await_shipment_stays_paid
      order = paid_order_for(customer, listing(seller))

      RollUpOrderStatus.new.call(order: order)

      assert_equal Domain::Orders::OrderStatus::PAID, order.reload.status
    end

    def test_one_shipped_fulfillment_of_two_partially_ships_the_order
      order = paid_order_for(customer, listing(seller("Blue Kiln Studio")), listing(seller("Rye Press")))
      order.fulfillments.first.update!(status: Domain::Orders::FulfillmentStatus::SHIPPED)

      RollUpOrderStatus.new.call(order: order)

      assert_equal Domain::Orders::OrderStatus::PARTIALLY_SHIPPED, order.reload.status
    end

    def test_every_fulfillment_delivered_delivers_the_order
      order = paid_order_for(customer, listing(seller))
      order.fulfillments.sole.update!(status: Domain::Orders::FulfillmentStatus::DELIVERED)

      RollUpOrderStatus.new.call(order: order)

      assert_equal Domain::Orders::OrderStatus::DELIVERED, order.reload.status
    end
  end
end

File: prototype/php/src/app/Actions/Orders/RollUpOrderStatusTest.php
<?php

declare(strict_types=1);

namespace App\Actions\Orders;

use App\Domain\Orders\FulfillmentStatus;
use App\Domain\Orders\OrderStatus;

it('is partially shipped when a paid order has one of two fulfillments shipped', function (): void {
    $order = $this->paidOrderWithTwoSellers();
    $order->fulfillments()->orderBy('id')->first()->update(['status' => FulfillmentStatus::Shipped]);

    $order = app(RollUpOrderStatus::class)($order->load('fulfillments'));

    expect($order->status)->toBe(OrderStatus::PartiallyShipped);
});

it('is shipped when every fulfillment has shipped', function (): void {
    $order = $this->paidOrderWithTwoSellers();
    $order->fulfillments()->update(['status' => FulfillmentStatus::Shipped]);

    $order = app(RollUpOrderStatus::class)($order->load('fulfillments'));

    expect($order->status)->toBe(OrderStatus::Shipped);
});

it('is delivered when every fulfillment has been delivered', function (): void {
    $order = $this->paidOrderWithTwoSellers();
    $order->fulfillments()->update(['status' => FulfillmentStatus::Shipped]);
    app(RollUpOrderStatus::class)($order->load('fulfillments'));
    $order->fulfillments()->update(['status' => FulfillmentStatus::Delivered]);

    $order = app(RollUpOrderStatus::class)($order->load('fulfillments'));

    expect($order->status)->toBe(OrderStatus::Delivered);
});
